### PR TITLE
Add `Options::additional_headers` and `subprotocols`

### PR DESCRIPTION
**Pull request by Its-Just-Nans**
Hello,

In the native app, the Websocket protocol can lead to error because it doesn't act like the web interface.

When we are using the WebSocket API (in the browser), it add automatically some headers - like the `Origin` header.

In fact this header is mandatory, as explained in the RFC :

- <https://datatracker.ietf.org/doc/html/rfc6455>

```text
8.   The request MUST include a header field with the name |Origin|
        [[RFC6454](https://datatracker.ietf.org/doc/html/rfc6454)] if the request is coming from a browser client.  If
        the connection is from a non-browser client, the request MAY
        include this header field if the semantics of that client match
        the use-case described here for browser clients.  The value of
        this header field is the ASCII serialization of origin of the
        context in which the code establishing the connection is
        running.  See [[RFC6454](https://datatracker.ietf.org/doc/html/rfc6454)] for the details of how this header field
        value is constructed.
```

However, if we don't set the `Origin` header in the native app, the app "could" not act the same as in the browser

In this PR, I had the Origin header to the client.

Note that
- the PR to add a header in tungstenite is quite new : https://github.com/snapview/tungstenite-rs/pull/400
- a new version of tungstenite has not been released yet (so this PR is kind of a draft)

Thanks

### CODE REVIEW

File: ewebsock/Cargo.toml [hunk 1]
```diff
@@ -39,11 +39,12 @@ tokio = [
 
 [dependencies]
 document-features = "0.2"
+http = "1.0.0"
 log = "0.4"
 
 # native:
 [target.'cfg(not(target_arch = "wasm32"))'.dependencies]
-tungstenite = { version = ">=0.17, <=0.20" }
+tungstenite = { version = ">=0.17, <=0.21" }
```
Comment (emilk) [inline ewebsock/Cargo.toml:47]: this needs to be set to the version number of the first tungstenite release with the `with_header` function

Comment (Its-Just-Nans) [reply ewebsock/Cargo.toml:47]: Yes you're right !
I'm waiting for the new version. I'm tagging the PR as DRAFT to wait for it

**Comment by Its-Just-Nans** [comment]
When tungstenite `0.22.0` will be out, this PR should fix

- [custom header addition](https://github.com/rerun-io/ewebsock/pull/27#issue-2132142672)
- https://github.com/rerun-io/ewebsock/pull/31
- https://github.com/rerun-io/ewebsock/pull/11

Now waiting for the version bump

**Comment by Its-Just-Nans** [comment]
Hi

tungstenite is now at 0.22 [ https://img.shields.io/crates/v/tungstenite](https://crates.io/crates/tungstenite)

Note that when the 0.23 will be out, it could be more convenient to use it https://github.com/snapview/tungstenite-rs/pull/427

**Comment by Its-Just-Nans** [comment]
There migth be a problem with tokio-tungstenite right now https://github.com/snapview/tokio-tungstenite/issues/334

(that's breaking my local CI/CD on my repo https://github.com/Its-Just-Nans/ewebsock/actions/runs/9337202702)

**Comment by Its-Just-Nans** [comment]
tungstenite and tokio-tungstenite are now at 0.23 !

I think this is mergeable now (local CI/CD pass: https://github.com/Its-Just-Nans/ewebsock/actions/runs/9338012481)

File: echo_server/src/main.rs [hunk 2]
```diff
@@ -11,14 +10,19 @@ fn main() {
         spawn(move || {
             let mut websocket = tungstenite::accept(stream.unwrap()).unwrap();
             eprintln!("New client connected");
-            while let Ok(msg) = websocket.read_message() {
+            loop {
+                let msg = websocket.read().unwrap();
```
Comment (emilk) [inline echo_server/src/main.rs:14]:
```suggestion
            while let Ok(msg) = websocket.read() {
```

File: README.md [hunk 3]
```diff
@@ -13,27 +13,29 @@ This is a simple [WebSocket](https://en.wikipedia.org/wiki/WebSocket) library fo
 ## Usage
 
 ``` rust
-let (mut sender, receiver) = ewebsock::connect("ws://example.com").unwrap();
+let options = ewebsock::Options::default();
+// add header and subprotocol to options
```
Comment (emilk) [inline README.md:17]: This is a confusing comment

File: ewebsock/src/lib.rs [hunk 4]
```diff
@@ -133,12 +133,24 @@ pub struct Options {
     ///
     /// Ignored on Web.
     pub max_incoming_frame_size: usize,
+
+    /// Additional Request headers
+    pub additional_headers: Option<Vec<(String, String)>>,
+
+    /// Additional subprotocols
```
Comment (emilk) [inline ewebsock/src/lib.rs:140]: Can we add a link here to some page that describes what a subprotocol is?

File: ewebsock/src/lib.rs [hunk 5]
```diff
@@ -133,12 +133,24 @@ pub struct Options {
     ///
     /// Ignored on Web.
     pub max_incoming_frame_size: usize,
+
+    /// Additional Request headers
```
Comment (emilk) [inline ewebsock/src/lib.rs:137]: Are these normal HTTP headers?

File: ewebsock/src/lib.rs [hunk 6]
```diff
@@ -133,12 +133,24 @@ pub struct Options {
...
+    pub additional_headers: Option<Vec<(String, String)>>,
```
Comment (emilk) [inline ewebsock/src/lib.rs:138]: What is the difference between `None` and `Some(vec![])`?
If there is none, just remove the `Option`

File: ewebsock/src/lib.rs [hunk 7]
```diff
@@ -133,12 +133,24 @@ pub struct Options {
...
+
+    /// Additional subprotocols
+    pub subprotocols: Option<Vec<String>>,
+
+    /// Delay blocking in ms - default 10ms
+    pub delay_blocking: Option<std::time::Duration>,
```
Comment (emilk) [inline ewebsock/src/lib.rs:144]: The default would be more apparent if you removed the `Option` and just set the 10 ms in the `default()`

File: ewebsock/src/native_tungstenite.rs [hunk 8]
```diff
@@ -165,15 +189,25 @@ pub fn ws_connect_blocking(
...
-        match tungstenite::client::connect_with_config(url, Some(config), max_redirects) {
-            Ok(result) => result,
-            Err(err) => {
-                return Err(format!("Connect: {err}"));
-            }
-        };
+    let uri: tungstenite::http::Uri = match url.parse() {
+        Ok(uri) => uri,
+        Err(err) => return Err(format!("Failed to parse URI: {err}")),
+    };
```
Comment (emilk) [inline ewebsock/src/native_tungstenite.rs:200]: nit: we can use `map_err` here and elsewhere:

```rs
let uri: tungstenite::http::Uri = url.parse().map_err(|err| format!("Failed to parse URI: {err}"))?;
```

**Review by emilk — COMMENTED** [review]
Thanks for the PR, but this PR now does three very different things:

A) replaces shell scripts with trunk
B) updates to tungstenite to 0.23
C) add default headers to the options

I would be thankful if you could break it into three PRs because
* it would be a lot easier to review
* it would be easier to revert one thing if we regret it
* it would be easier to git-bisect bugs in the future

**Comment by Its-Just-Nans** [comment]
okay, here my next steps:

- fix the current pr according to review
- create 3 PRs

File: ewebsock/src/lib.rs (hunk 5 above)
Comment (Its-Just-Nans) [reply ewebsock/src/lib.rs:137]: Yes, but they can also be custom header

**Review by emilk — APPROVED** [review]

File: ewebsock/src/lib.rs [hunk 9]
```diff
@@ -134,6 +134,14 @@ pub struct Options {
     /// Ignored on Web.
     pub max_incoming_frame_size: usize,
 
+    /// Additional Request headers
+    pub additional_headers: Vec<(String, String)>,
+
+    /// Additional subprotocols -
+    /// <https://www.iana.org/assignments/websocket/websocket.xml>
+    /// <https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API/Writing_WebSocket_servers#miscellaneous>
+    pub subprotocols: Vec<String>,
```
Comment (emilk) [inline ewebsock/src/lib.rs:143]:
```suggestion
    /// Additional Request headers.
    ///
    /// Currently only supported on native.
    pub additional_headers: Vec<(String, String)>,

    /// Additional subprotocols.
    ///
    /// <https://www.iana.org/assignments/websocket/websocket.xml>
    /// <https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API/Writing_WebSocket_servers#miscellaneous>
    ///
    /// Currently only supported on native.
    pub subprotocols: Vec<String>,
```

**Pull request merged**